<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Extending cooccurNet</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Extending cooccurNet</h1>



<p>This vignette documents the official extension mechanism provided in cooccurNet 0.1.4. Most of this information is available scattered throughout the R documentation. This vignette brings it all together in one place.</p>
<p>As you read this document, you’ll learn to what the cooccurNet is, how to run the methods, and how to interpret, plot and report outputs.</p>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p><strong>cooccurNet</strong> is an R package including functionalities of construction and analysis of residue (amino acid, nucleotide, SNP and so on) co-occurrence network. Besides, a new method for measuring residues coevolution, defined as <strong>residue co-occurrence score (RCOS)</strong>, was proposed and implemented in it based on the co-occurrence network.</p>
<p>The co-occurrence network method was originally developed to capture the nucleotide co-occurrence pattern over the entire genome of human influenza H3N2 viruses [1]. It could effectively capture the evolutionary antigenic features of H3N2 virus at the whole-genome level and accurately describe the complex evolutionary patterns between individual gene segments. Besides influenza viruses, the co-occurrence network method was also used to Zaire ebolavirus (EBOV). In this work [2], we found that the features of co-occurrence networks (edges and nodes) built over the EBOV genome could accurately reflect the case fatality rate (CFR) of EBOV, which provided a novel and rapid way of assessing the CFR of EBOV even at the very beginning of the outbreak once its genome sequence is available. This suggests the potential usage of the co-occurrence network method in modeling the evolution of genomes.</p>
<p>As we know, the genomes contain all the genetic information of an organism, the evolution of which is very complex. Currently, there is a lack of computational models for capturing the complex evolutionary features over the entire genome. Coevolution is prevalent in biology and it is an important driving force in evolution. The co-occurrence network method could capture the coevolution linkage between residues within the genome. It is unique in that it could transform a genome into a residue co-occurrence network, the features of which, such as the nodes, edges, connectivity and so on, may reflect important features of organisms, such as the antigenicity [1] and CFR [2] in the examples mentioned above.</p>
<p>Actually, the co-occurrence network method is a general method which could be used to capture the co-occurrence pattern within samples. Besides genomes, it could be directly used in gene, protein, proteome, and so on. Due to a lack of a user-friendly software, the method is rarely used in other data type except for genomes. Based on the package “cooccurNet”, we believe the method could be widely used in biology.</p>
<p>The <strong>RCOS</strong> method was newly proposed in the package. It is a method for measuring the association between two discrete variables. It could be used to measure the extent of residue coevolution in protein/DNA/RNA/SNP. Residue coevolution reflects the functional and structural constraints within a protein or gene. It could help for protein structure or function prediction [3-6]. Many computational methods have been developed in measure residue coevolution. Our tests show that the RCOS method achieved a similar performance with a state-of-the-art method PSICOV on a benchmark dataset [7]. Therefore, we believe the <strong>RCOS</strong> method could become an important method in measuring residue coevolution.</p>
<p>There are two important parameters for both the methods of co-occurrence network and RCOS in the package “cooccurNet”. The first parameter, called as “conservativeFilter”, is used to filter the highly conservative sites for which the ratio of some residue is larger than the “conservativeFilter”. By default, it is set to be 0.95 for all data types. The second parameter, called as “cooccurFilter”, is used to determine whether a pair of residues has co-occurrence. According to the methods of co-occurrence network and RCOS, the smaller the “cooccurFilter” is, more pairs of residues would have co-occurrence and more false positives would be observed. To investigate the influence of this cutoff on the RCOS method, we used the RCOS method with cutoffs ranging from 0.8 to 1 to identify the structural constraints between residues based on the benchmark dataset derived from Jones’s work [5]. As shown in Table 1 below, as the cutoff increased, the number of residue pairs with significant co-occurrence decreased, while the ratio of residue pairs in contact among the top L/10, L/5, L/2 and L ranked pairs of residues increased. When the cutoff was set to be 1, only a few residue pairs with significant co-occurrence could be identified for most Pfam families. For balance of these two measures, we defined this cutoff to be 0.9 in default for the protein in the package. While for the other data types, such as DNA, RNA and SNP, “cooccurFilter” was set to be 1 in default. Both the parameters “conservativeFilter” and “cooccurFilter” could be re-set in the package.</p>
<p><strong>Table 1</strong>: Performance of the RCOS method with the cutoff of “cooccurFilter” ranging from 0.8 to 1. For each value of the cutoff, it lists the average number of pairs of residues with significant co-occurrence (p-value &lt; 0.05) identified by the RCOS method, and the ratios of residue pairs in contact among the top L/10, L/5, L/2 and L ranked pairs of residues for 150 Pfam families by the method of RCOS. “*“, not applicable due to fewer than L/10 pairs of residues with significant co-occurrence were identified.</p>
<table style="width:117%;">
<colgroup>
<col width="22%"></col>
<col width="55%"></col>
<col width="9%"></col>
<col width="9%"></col>
<col width="9%"></col>
<col width="9%"></col>
</colgroup>
<thead>
<tr class="header">
<th align="left">Co-occurFilter</th>
<th align="left">Average number of co-occurrence pairs</th>
<th align="left">L/10</th>
<th align="left">L/5</th>
<th align="left">L/2</th>
<th align="left">L</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">0.8</td>
<td align="left">221</td>
<td align="left">0.59</td>
<td align="left">0.56</td>
<td align="left">0.47</td>
<td align="left">0.36</td>
</tr>
<tr class="even">
<td align="left">0.85</td>
<td align="left">153</td>
<td align="left">0.66</td>
<td align="left">0.62</td>
<td align="left">0.49</td>
<td align="left">0.37</td>
</tr>
<tr class="odd">
<td align="left">0.9</td>
<td align="left">95</td>
<td align="left">0.76</td>
<td align="left">0.69</td>
<td align="left">0.54</td>
<td align="left">0.42</td>
</tr>
<tr class="even">
<td align="left">0.95</td>
<td align="left">51</td>
<td align="left">0.85</td>
<td align="left">0.75</td>
<td align="left">0.57</td>
<td align="left">0.45</td>
</tr>
<tr class="odd">
<td align="left">1</td>
<td align="left">6</td>
<td align="left">-*</td>
<td align="left">-</td>
<td align="left">-</td>
<td align="left">-</td>
</tr>
</tbody>
</table>
<p><strong>References:</strong></p>
<p>[1] Du X, Wang Z, Wu A, et al. Networks of genomic co-occurrence capture characteristics of human influenza A (H3N2) evolution[J]. Genome research, 2008, 18(1): 178-187.</p>
<p>[2] Deng L, Liu M, Hua S, et al. Network of co-mutations in Ebola virus genome predicts the disease lethality[J]. Cell research, 2015, 25(6): 753.</p>
<p>[3] Lee B C, Park K, Kim D. Analysis of the residue–residue coevolution network and the functionally important residues in proteins[J]. Proteins: Structure, Function, and Bioinformatics, 2008, 72(3): 863-872.</p>
<p>[4] Aurora R, Donlin M J, Cannon N A, et al. Genome-wide hepatitis C virus amino acid covariance networks can predict response to antiviral therapy in humans[J]. The Journal of clinical investigation, 2009, 119(1): 225-236.</p>
<p>[5] Jones D T, Buchan D W A, Cozzetto D, et al. PSICOV: precise structural contact prediction using sparse inverse covariance estimation on large multiple sequence alignments[J]. Bioinformatics, 2012, 28(2): 184-190.</p>
<p>[6] Wang S, Li W, Zhang R, et al. CoinFold: a web server for protein contact prediction and contact-assisted protein folding[J]. Nucleic acids research, 2016: gkw307.</p>
<p>[7] Zou Y, Wu Z, et al. cooccurNet: an R package for co-occurrence network construction and analysis. Bioinformatics (In submission).</p>
</div>
<div id="preliminaries" class="section level2">
<h2>Preliminaries</h2>
<div id="osmemory" class="section level3">
<h3>OS/Memory</h3>
<ul>
<li><strong>cooccurNet</strong> compiles and runs on a wide variety of UNIX platforms, Windows and MacOS.</li>
<li>System memory should <strong>≥ 2Gb</strong>. See section “Optimization” for more details.</li>
</ul>
</div>
<div id="how-to-get-help" class="section level3">
<h3>How to get help</h3>
<p>Most questions about cooccurNet will hopefully be answered by the documentation or references.</p>
<ul>
<li><p><strong>cooccurNet</strong> is updated frequently. Once you have installed cooccurNet, You can get help from the change-log.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">logs &lt;-<span class="st"> </span><span class="kw">changeLog</span>(<span class="dt">n=</span><span class="dv">20</span>)
<span class="co">#&gt; </span>
logs
<span class="co">#&gt;  [1] &quot;2016-12-16 Yuanqiang Zou  &lt;jerrytsou2001@gmail.com&gt;&quot;           </span>
<span class="co">#&gt;  [2] &quot;&quot;                                                              </span>
<span class="co">#&gt;  [3] &quot;\tDocument synchronization and test outputs.&quot;                  </span>
<span class="co">#&gt;  [4] &quot;&quot;                                                              </span>
<span class="co">#&gt;  [5] &quot;2017-01-10 Yuanqiang Zou  &lt;jerrytsou2001@gmail.com&gt;&quot;           </span>
<span class="co">#&gt;  [6] &quot;&quot;                                                              </span>
<span class="co">#&gt;  [7] &quot;\tModifiy documentaion and import a new package \&quot;doParallel\&quot;&quot;</span>
<span class="co">#&gt;  [8] &quot;&quot;                                                              </span>
<span class="co">#&gt;  [9] &quot;2017-01-27 Yuanqiang Zou  &lt;jerrytsou2001@gmail.com&gt;&quot;           </span>
<span class="co">#&gt; [10] &quot;&quot;                                                              </span>
<span class="co">#&gt; [11] &quot;\tOptimized the fuction \&quot;toigraph\&quot;&quot;</span></code></pre></div></li>
<li><p>If you’ve run into a problem which could not be addressed by the documentation, or if you’ve found a conflict between the documentation and software itself, then please write to us: Yuanqiang Zou <a href="mailto:jerrytsou2001@gmail.com">jerrytsou2001@gmail.com</a> or Yousong Peng <a href="mailto:pys2013@hnu.edu.cn">pys2013@hnu.edu.cn</a>.</p></li>
</ul>
</div>
</div>
<div id="quick-start" class="section level2">
<h2>Quick Start</h2>
<p>In this section, you’ll learn to how to run the methods by using the sample data.</p>
<div id="get-sample-sequence-file" class="section level3">
<h3>Get sample sequence file</h3>
<p>The <strong>parameter</strong> <code>dataType</code> could be ‘DNA’, ‘protein’, ‘SNP’, ‘RNA’. It’s ‘protein’ by default.</p>
<div id="get-protein-sample-data-file-name-with-full-path" class="section level4">
<h4>Get protein sample data file name with full path</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataFile=<span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>)
<span class="co">#&gt; </span>
dataFile
<span class="co">#&gt; [1] &quot;C:/Users/Jerry/AppData/Local/Temp/Rtmpgvbts6/Rinst146815772caf/cooccurNet/extdata/protein&quot;</span></code></pre></div>
</div>
<div id="get-dna-sample-data-file-name-with-full-path" class="section level4">
<h4>Get DNA sample data file name with full path</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataFile=<span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;DNA&quot;</span>)
<span class="co">#&gt; </span>
dataFile
<span class="co">#&gt; [1] &quot;C:/Users/Jerry/AppData/Local/Temp/Rtmpgvbts6/Rinst146815772caf/cooccurNet/extdata/DNA&quot;</span></code></pre></div>
</div>
<div id="get-snp-sample-data-file-name-with-full-path" class="section level4">
<h4>Get SNP sample data file name with full path</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataFile=<span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;SNP&quot;</span>)
<span class="co">#&gt; </span>
dataFile
<span class="co">#&gt; [1] &quot;C:/Users/Jerry/AppData/Local/Temp/Rtmpgvbts6/Rinst146815772caf/cooccurNet/extdata/SNP&quot;</span></code></pre></div>
</div>
<div id="get-rna-sample-data-file-name-with-full-path" class="section level4">
<h4>Get RNA sample data file name with full path</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataFile=<span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;RNA&quot;</span>)
<span class="co">#&gt; </span>
dataFile
<span class="co">#&gt; [1] &quot;C:/Users/Jerry/AppData/Local/Temp/Rtmpgvbts6/Rinst146815772caf/cooccurNet/extdata/RNA&quot;</span></code></pre></div>
</div>
<div id="get-sample-rcos-protein-sequence-file-names-sample-human-h3n2-ha-protein-sequence-file-names-with-full-path" class="section level4">
<h4>Get sample RCOS protein sequence file names (sample human H3N2 HA protein sequence file names with full path)</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataFiles=<span class="kw">getexample_forRCOS</span>()
<span class="co">#&gt; </span>
dataFiles
<span class="co">#&gt; [1] &quot;C:/Users/Jerry/AppData/Local/Temp/Rtmpgvbts6/Rinst146815772caf/cooccurNet/extdata/RCOS/HA1protein_humanH3N2&quot;</span></code></pre></div>
</div>
</div>
<div id="read-fasta-format-sequence-file" class="section level3">
<h3>Read FASTA format sequence file</h3>
<p>To save the RAM usage by the package, we have designed the method of using prime number to stand for the nucleotides or amino acid in the data inputted. As is known to us, in the R environment an integer vector uses 4 bytes per number, while a character vector uses 8 bytes per character. Using prime number to stand for character could save half of RAM usage for storing the data inputted. Besides, when calculating the frequency of a pair of characters (nucleotide or amino acid), it is necessary to concatenate the characters in R. After replacing the characters with prime numbers, for a pair of prime numbers A and B, we found that unique number could be derived using the following formula for prime numbers (starting from 2):</p>
<pre><code>           (round(A/B,5)+B)*100000                    (1)
           </code></pre>
<p>The resulting values have a one-to-one correspondence with the pairs of nucleotide or amino acid. Therefore, the concentration of character vectors could be transformed to algebraic operation of prime number vectors, which could significantly reduce the time consumed in this process.</p>
<p>The <strong>parameter</strong> <code>dataFile</code> is a FASTA data file name with full path.</p>
<div id="read-protein-sample-data" class="section level4">
<h4>Read protein sample data</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data =<span class="st"> </span><span class="kw">readseq</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>)
<span class="co">#&gt; Reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
data$original[<span class="dv">1</span>:<span class="dv">10</span>,<span class="dv">1</span>:<span class="dv">10</span>]
<span class="co">#&gt;       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span>
<span class="co">#&gt;  [1,] &quot;Q&quot;  &quot;K&quot;  &quot;I&quot;  &quot;P&quot;  &quot;G&quot;  &quot;N&quot;  &quot;D&quot;  &quot;N&quot;  &quot;S&quot;  &quot;T&quot;  </span>
<span class="co">#&gt;  [2,] &quot;Q&quot;  &quot;K&quot;  &quot;L&quot;  &quot;P&quot;  &quot;G&quot;  &quot;N&quot;  &quot;D&quot;  &quot;N&quot;  &quot;S&quot;  &quot;T&quot;  </span>
<span class="co">#&gt;  [3,] &quot;Q&quot;  &quot;K&quot;  &quot;I&quot;  &quot;P&quot;  &quot;G&quot;  &quot;N&quot;  &quot;D&quot;  &quot;N&quot;  &quot;S&quot;  &quot;T&quot;  </span>
<span class="co">#&gt;  [4,] &quot;Q&quot;  &quot;K&quot;  &quot;I&quot;  &quot;P&quot;  &quot;G&quot;  &quot;N&quot;  &quot;D&quot;  &quot;N&quot;  &quot;S&quot;  &quot;T&quot;  </span>
<span class="co">#&gt;  [5,] &quot;Q&quot;  &quot;K&quot;  &quot;L&quot;  &quot;P&quot;  &quot;G&quot;  &quot;N&quot;  &quot;D&quot;  &quot;N&quot;  &quot;S&quot;  &quot;T&quot;  </span>
<span class="co">#&gt;  [6,] &quot;Q&quot;  &quot;K&quot;  &quot;V&quot;  &quot;P&quot;  &quot;G&quot;  &quot;N&quot;  &quot;D&quot;  &quot;N&quot;  &quot;S&quot;  &quot;T&quot;  </span>
<span class="co">#&gt;  [7,] &quot;Q&quot;  &quot;K&quot;  &quot;L&quot;  &quot;P&quot;  &quot;G&quot;  &quot;N&quot;  &quot;D&quot;  &quot;N&quot;  &quot;S&quot;  &quot;T&quot;  </span>
<span class="co">#&gt;  [8,] &quot;Q&quot;  &quot;K&quot;  &quot;I&quot;  &quot;P&quot;  &quot;G&quot;  &quot;N&quot;  &quot;D&quot;  &quot;N&quot;  &quot;S&quot;  &quot;T&quot;  </span>
<span class="co">#&gt;  [9,] &quot;Q&quot;  &quot;K&quot;  &quot;L&quot;  &quot;P&quot;  &quot;G&quot;  &quot;N&quot;  &quot;D&quot;  &quot;N&quot;  &quot;S&quot;  &quot;T&quot;  </span>
<span class="co">#&gt; [10,] &quot;Q&quot;  &quot;K&quot;  &quot;L&quot;  &quot;P&quot;  &quot;G&quot;  &quot;N&quot;  &quot;D&quot;  &quot;N&quot;  &quot;S&quot;  &quot;T&quot;</span>
data$matrix[<span class="dv">1</span>:<span class="dv">10</span>,<span class="dv">1</span>:<span class="dv">10</span>]
<span class="co">#&gt;        1  2  3  4  5  6 7  8  9 10</span>
<span class="co">#&gt;  [1,] 43 23 19 41 13 37 5 37 53 59</span>
<span class="co">#&gt;  [2,] 43 23 29 41 13 37 5 37 53 59</span>
<span class="co">#&gt;  [3,] 43 23 19 41 13 37 5 37 53 59</span>
<span class="co">#&gt;  [4,] 43 23 19 41 13 37 5 37 53 59</span>
<span class="co">#&gt;  [5,] 43 23 29 41 13 37 5 37 53 59</span>
<span class="co">#&gt;  [6,] 43 23 61 41 13 37 5 37 53 59</span>
<span class="co">#&gt;  [7,] 43 23 29 41 13 37 5 37 53 59</span>
<span class="co">#&gt;  [8,] 43 23 19 41 13 37 5 37 53 59</span>
<span class="co">#&gt;  [9,] 43 23 29 41 13 37 5 37 53 59</span>
<span class="co">#&gt; [10,] 43 23 29 41 13 37 5 37 53 59</span></code></pre></div>
</div>
<div id="read-dna-sample-data" class="section level4">
<h4>Read DNA sample data</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data =<span class="st"> </span><span class="kw">readseq</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;DNA&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;DNA&quot;</span>)
<span class="co">#&gt; Reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (213,987)</span>
<span class="co">#&gt; </span>
data$original[<span class="dv">1</span>:<span class="dv">10</span>,<span class="dv">1</span>:<span class="dv">10</span>]
<span class="co">#&gt;       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span>
<span class="co">#&gt;  [1,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;T&quot;  &quot;T&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [2,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;C&quot;  &quot;T&quot;  &quot;T&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [3,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;T&quot;  &quot;T&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [4,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;T&quot;  &quot;T&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [5,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;T&quot;  &quot;T&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [6,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;T&quot;  &quot;T&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [7,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;T&quot;  &quot;T&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [8,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;T&quot;  &quot;T&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [9,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;T&quot;  &quot;T&quot;  &quot;C&quot;  </span>
<span class="co">#&gt; [10,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;T&quot;  &quot;T&quot;  &quot;C&quot;</span>
data$matrix[<span class="dv">1</span>:<span class="dv">10</span>,<span class="dv">1</span>:<span class="dv">10</span>]
<span class="co">#&gt;       1 2 3 4 5 6 7 8 9 10</span>
<span class="co">#&gt;  [1,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [2,] 3 2 2 2 2 3 3 7 7  3</span>
<span class="co">#&gt;  [3,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [4,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [5,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [6,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [7,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [8,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [9,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt; [10,] 3 2 2 2 2 2 3 7 7  3</span></code></pre></div>
</div>
<div id="read-snp-sample-data" class="section level4">
<h4>Read SNP sample data</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data =<span class="st"> </span><span class="kw">readseq</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;SNP&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;SNP&quot;</span>)
<span class="co">#&gt; Reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (100,50)</span>
<span class="co">#&gt; </span>
data$original[<span class="dv">1</span>:<span class="dv">10</span>,<span class="dv">1</span>:<span class="dv">10</span>]
<span class="co">#&gt;       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span>
<span class="co">#&gt;  [1,] &quot;1&quot;  &quot;0&quot;  &quot;1&quot;  &quot;1&quot;  &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  </span>
<span class="co">#&gt;  [2,] &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot;  &quot;1&quot;  &quot;0&quot;  &quot;1&quot;  &quot;1&quot;  &quot;0&quot;  </span>
<span class="co">#&gt;  [3,] &quot;1&quot;  &quot;0&quot;  &quot;1&quot;  &quot;1&quot;  &quot;1&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  </span>
<span class="co">#&gt;  [4,] &quot;1&quot;  &quot;1&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot;  &quot;1&quot;  &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  &quot;1&quot;  </span>
<span class="co">#&gt;  [5,] &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  </span>
<span class="co">#&gt;  [6,] &quot;0&quot;  &quot;0&quot;  &quot;1&quot;  &quot;1&quot;  &quot;1&quot;  &quot;1&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  </span>
<span class="co">#&gt;  [7,] &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot;  </span>
<span class="co">#&gt;  [8,] &quot;1&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  </span>
<span class="co">#&gt;  [9,] &quot;1&quot;  &quot;1&quot;  &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  &quot;1&quot;  &quot;1&quot;  &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  </span>
<span class="co">#&gt; [10,] &quot;0&quot;  &quot;1&quot;  &quot;1&quot;  &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  &quot;0&quot;  &quot;1&quot;  &quot;0&quot;  &quot;1&quot;</span>
data$matrix[<span class="dv">1</span>:<span class="dv">10</span>,<span class="dv">1</span>:<span class="dv">10</span>]
<span class="co">#&gt;       1 2 3 4 5 6 7 8 9 10</span>
<span class="co">#&gt;  [1,] 3 2 3 3 2 3 2 2 2  2</span>
<span class="co">#&gt;  [2,] 2 3 2 2 3 3 2 3 3  2</span>
<span class="co">#&gt;  [3,] 3 2 3 3 3 2 2 2 3  2</span>
<span class="co">#&gt;  [4,] 3 3 2 2 3 3 2 3 2  3</span>
<span class="co">#&gt;  [5,] 2 2 2 3 2 3 2 2 3  2</span>
<span class="co">#&gt;  [6,] 2 2 3 3 3 3 2 2 2  2</span>
<span class="co">#&gt;  [7,] 2 3 2 2 2 2 2 2 2  3</span>
<span class="co">#&gt;  [8,] 3 2 2 2 2 2 2 2 3  2</span>
<span class="co">#&gt;  [9,] 3 3 2 3 2 3 3 2 3  2</span>
<span class="co">#&gt; [10,] 2 3 3 2 3 2 2 3 2  3</span></code></pre></div>
</div>
<div id="read-rna-sample-data" class="section level4">
<h4>Read RNA sample data</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data =<span class="st"> </span><span class="kw">readseq</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;RNA&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;RNA&quot;</span>)
<span class="co">#&gt; Reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (45,987)</span>
<span class="co">#&gt; </span>
data$original[<span class="dv">1</span>:<span class="dv">10</span>,<span class="dv">1</span>:<span class="dv">10</span>]
<span class="co">#&gt;       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span>
<span class="co">#&gt;  [1,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;U&quot;  &quot;U&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [2,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;C&quot;  &quot;U&quot;  &quot;U&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [3,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;U&quot;  &quot;U&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [4,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;U&quot;  &quot;U&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [5,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;U&quot;  &quot;U&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [6,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;U&quot;  &quot;U&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [7,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;U&quot;  &quot;U&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [8,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;U&quot;  &quot;U&quot;  &quot;C&quot;  </span>
<span class="co">#&gt;  [9,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;U&quot;  &quot;U&quot;  &quot;C&quot;  </span>
<span class="co">#&gt; [10,] &quot;C&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;A&quot;  &quot;C&quot;  &quot;U&quot;  &quot;U&quot;  &quot;C&quot;</span>
data$matrix[<span class="dv">1</span>:<span class="dv">10</span>,<span class="dv">1</span>:<span class="dv">10</span>]
<span class="co">#&gt;       1 2 3 4 5 6 7 8 9 10</span>
<span class="co">#&gt;  [1,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [2,] 3 2 2 2 2 3 3 7 7  3</span>
<span class="co">#&gt;  [3,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [4,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [5,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [6,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [7,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [8,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt;  [9,] 3 2 2 2 2 2 3 7 7  3</span>
<span class="co">#&gt; [10,] 3 2 2 2 2 2 3 7 7  3</span></code></pre></div>
</div>
</div>
<div id="pre-process-the-data-inputted" class="section level3">
<h3>Pre-process the data inputted</h3>
<p>The <strong>parameter</strong> <code>conservativeFilter</code> is used to filter the highly conservative columns which the ratio of some residue is larger than the conservationFilter. By default, it is set to be 0.95.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data =<span class="st"> </span><span class="kw">readseq</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>)
<span class="co">#&gt; Reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
data_process =<span class="st"> </span><span class="kw">pprocess</span>(<span class="dt">data=</span>data,<span class="dt">conservativeFilter=</span><span class="fl">0.95</span>)
<span class="co">#&gt; Pre-processing......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating bigram frequency ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Pre-process......completed. (10,42)</span>
<span class="co">#&gt; </span></code></pre></div>
</div>
<div id="generate-co-occurrence-network-and-save-it-into-networkfile" class="section level3">
<h3>Generate co-occurrence network and save it into ‘networkFile’</h3>
<p>For each sequence, a residue co-occurrence network would be generated. It measures the covariations between residues in the sequence. The features and properties of the network may capture important features of species, such as the antigenicity and CFR mentioned above in the section of introduction.</p>
<p>The <strong>parameter</strong> <code>cooccurFilter</code> is used to determine whether a pair of residues have perfect co-occurrence. By default, it is set to be 0.90 for the data type of protein, while it is set to be 1 for the data type of DNA, RNA, SNP and so on.</p>
<p>The <strong>parameter</strong> <code>networkFile</code> is a file name with full path for storing the co-occurrence network for each row. In default, it is set to be ‘cooccurNetwork’, which is in your current workspace (<code>getwd()</code>)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cooccurNetwork =<span class="st"> </span><span class="kw">gencooccur</span>(<span class="dt">data=</span>data_process, <span class="dt">cooccurFilter=</span><span class="fl">0.9</span>, <span class="dt">networkFile=</span><span class="st">'cooccurNetwork'</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; 1. calculating networks ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2. creating network file (cooccurNetwork) ......completed</span>
<span class="co">#&gt; </span>
workspace =<span class="st"> </span><span class="kw">getwd</span>()
workspace
<span class="co">#&gt; [1] &quot;C:/Users/Jerry/AppData/Local/Temp/Rtmpgvbts6/Rbuild146832c62740/cooccurNet/vignettes&quot;</span>
<span class="kw">readLines</span>(cooccurNetwork$networkFile)
<span class="co">#&gt;  [1] &quot;EPI823725 3-144 3-159 3-311 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                </span>
<span class="co">#&gt;  [2] &quot;EPI775768 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-326 186-219 186-278 208-238 208-261 219-278 238-261 267-312&quot;                                                                                  </span>
<span class="co">#&gt;  [3] &quot;EPI814365 3-144 3-159 3-311 25-121 25-216 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 142-197 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                          </span>
<span class="co">#&gt;  [4] &quot;EPI787403 3-144 3-159 3-311 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 142-197 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                        </span>
<span class="co">#&gt;  [5] &quot;EPI787049 25-208 25-238 25-261 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-326 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                     </span>
<span class="co">#&gt;  [6] &quot;EPI775712 3-140 3-158 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-225 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                              </span>
<span class="co">#&gt;  [7] &quot;EPI777683 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 159-225 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                                                  </span>
<span class="co">#&gt;  [8] &quot;EPI787395 3-144 3-159 3-311 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                </span>
<span class="co">#&gt;  [9] &quot;EPI731085 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-326 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                                          </span>
<span class="co">#&gt; [10] &quot;EPI748184 43-45 43-48 43-144 43-186 43-219 43-267 43-278 45-48 45-144 45-186 45-219 45-267 45-278 48-144 48-186 48-219 48-267 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 144-186 144-219 144-267 144-278 159-225 186-219 186-267 186-278 208-238 208-261 219-267 219-278 238-261 267-278&quot;</span></code></pre></div>
</div>
<div id="generate-co-occurrence-network-file-by-one-command" class="section level3">
<h3>Generate co-occurrence network file by one command</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cooccurNetwork =<span class="st"> </span><span class="kw">coocnet</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>,<span class="dt">conservativeFilter=</span><span class="fl">0.95</span>, <span class="dt">cooccurFilter=</span><span class="fl">0.9</span>, <span class="dt">networkFile=</span><span class="st">'cooccurNetwork'</span>)
<span class="co">#&gt; reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating bigram frequency ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......completed. (10,42)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 1. calculating networks ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2. creating network file (cooccurNetwork) ......completed</span>
<span class="co">#&gt; </span>
<span class="kw">readLines</span>(cooccurNetwork$networkFile)
<span class="co">#&gt;  [1] &quot;EPI823725 3-144 3-159 3-311 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                </span>
<span class="co">#&gt;  [2] &quot;EPI775768 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-326 186-219 186-278 208-238 208-261 219-278 238-261 267-312&quot;                                                                                  </span>
<span class="co">#&gt;  [3] &quot;EPI814365 3-144 3-159 3-311 25-121 25-216 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 142-197 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                          </span>
<span class="co">#&gt;  [4] &quot;EPI787403 3-144 3-159 3-311 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 142-197 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                        </span>
<span class="co">#&gt;  [5] &quot;EPI787049 25-208 25-238 25-261 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-326 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                     </span>
<span class="co">#&gt;  [6] &quot;EPI775712 3-140 3-158 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-225 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                              </span>
<span class="co">#&gt;  [7] &quot;EPI777683 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 159-225 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                                                  </span>
<span class="co">#&gt;  [8] &quot;EPI787395 3-144 3-159 3-311 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                </span>
<span class="co">#&gt;  [9] &quot;EPI731085 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-326 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                                          </span>
<span class="co">#&gt; [10] &quot;EPI748184 43-45 43-48 43-144 43-186 43-219 43-267 43-278 45-48 45-144 45-186 45-219 45-267 45-278 48-144 48-186 48-219 48-267 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 144-186 144-219 144-267 144-278 159-225 186-219 186-267 186-278 208-238 208-261 219-267 219-278 238-261 267-278&quot;</span></code></pre></div>
</div>
<div id="generate-co-occurrence-network-module-file-by-one-command" class="section level3">
<h3>Generate co-occurrence network module file by one command</h3>
<p>This would calculate all the modules (sub-graphs) included in each residue co-occurrence network.</p>
<p>The <strong>parameter</strong> <code>module</code> is a boolean flag to check whether the modules in each network of the networkFile would be calculated.</p>
<p>The <strong>parameter</strong> <code>moduleFile</code> is a file name with full path for storing the modules for co-occurrence network. It’s ‘cooccurNetworkModule’ by default.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cooccurNetwork  =<span class="st"> </span><span class="kw">coocnet</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>,<span class="dt">conservativeFilter=</span><span class="fl">0.95</span>, <span class="dt">cooccurFilter=</span><span class="fl">0.9</span>, <span class="dt">module =</span> <span class="ot">TRUE</span>, <span class="dt">moduleFile=</span><span class="st">'cooccurNetworkModule'</span>)
<span class="co">#&gt; reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating bigram frequency ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......completed. (10,42)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 1. calculating networks ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2. creating network file (cooccurNetwork) ......completed</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 3. creating module file (cooccurNetworkModule) ......completed</span>
<span class="co">#&gt; </span>
<span class="kw">readLines</span>(cooccurNetwork$moduleFile)
<span class="co">#&gt;  [1] &quot;EPI823725 (3,144,159,311) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (140,158) (208,238,261)&quot;              </span>
<span class="co">#&gt;  [2] &quot;EPI775768 (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (128,142) (140,158) (159,326) (208,238,261) (267,312)&quot;</span>
<span class="co">#&gt;  [3] &quot;EPI814365 (3,144,159,311) (25,121,216) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (140,158) (142,197) (208,238,261)&quot; </span>
<span class="co">#&gt;  [4] &quot;EPI787403 (3,144,159,311) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (140,158) (142,197) (208,238,261)&quot;    </span>
<span class="co">#&gt;  [5] &quot;EPI787049 (25,208,238,261) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (128,142) (140,158) (159,326)&quot;       </span>
<span class="co">#&gt;  [6] &quot;EPI775712 (3,140,158) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (128,142) (159,225) (208,238,261)&quot;        </span>
<span class="co">#&gt;  [7] &quot;EPI777683 (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (140,158) (159,225) (208,238,261)&quot;                    </span>
<span class="co">#&gt;  [8] &quot;EPI787395 (3,144,159,311) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (140,158) (208,238,261)&quot;              </span>
<span class="co">#&gt;  [9] &quot;EPI731085 (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (128,142) (140,158) (159,326) (208,238,261)&quot;          </span>
<span class="co">#&gt; [10] &quot;EPI748184 (43,45,48,144,186,219,267,278) (53,114,203) (62,259) (92,194) (121,216) (140,158) (159,225) (208,238,261)&quot;</span></code></pre></div>
</div>
<div id="generate-co-occurrence-network-property-file-by-one-command" class="section level3">
<h3>Generate co-occurrence network property file by one command</h3>
<p>The <strong>parameter</strong> <code>property</code> is a boolean flag to check whether the properties for each network of the networkFile, including the network diameter, connectivity, ConnectionEffcient and so on, would be calculated.</p>
<p>The <strong>parameter</strong> <code>propertyFile</code> is file name with full path storing the properties for each network of the networkFile. It’s ‘cooccurNetworkProperty’ by default.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cooccurNetwork  =<span class="st"> </span><span class="kw">coocnet</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>,<span class="dt">conservativeFilter=</span><span class="fl">0.95</span>, <span class="dt">cooccurFilter=</span><span class="fl">0.9</span>, <span class="dt">property =</span> <span class="ot">TRUE</span>, <span class="dt">propertyFile=</span><span class="st">'cooccurNetworkProperty'</span>)
<span class="co">#&gt; reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating bigram frequency ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......completed. (10,42)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 1. calculating networks ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2. creating network file (cooccurNetwork) ......completed</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 3. creating property file (cooccurNetworkProperty) ......completed</span>
<span class="co">#&gt; </span>
<span class="kw">readLines</span>(cooccurNetwork$propertyFile)[<span class="dv">1</span>:<span class="dv">5</span>]
<span class="co">#&gt; [1] &quot;name  Connectivity  Diameter  Radius  ConnectionEffcient&quot;</span>
<span class="co">#&gt; [2] &quot;EPI823725 0 1 1 1&quot;                                       </span>
<span class="co">#&gt; [3] &quot;EPI775768 0 1 1 1&quot;                                       </span>
<span class="co">#&gt; [4] &quot;EPI814365 0 1 1 1&quot;                                       </span>
<span class="co">#&gt; [5] &quot;EPI787403 0 1 1 1&quot;</span></code></pre></div>
</div>
<div id="generate-the-residue-co-occurence-file-by-one-command" class="section level3">
<h3>Generate the residue co-occurence file by one command</h3>
<p>This is to calculate the pairwise residue co-occurrence score (RCOS) for all pairs of positions within the sequences. To evaluate the statistical significance of RCOS, the data would be permutated N times (N=100 in default). This process would take a long time. The RCOS was calculated in the permutated data and sorted decreasingly. The rank of the RCOS derived from the original data in the RCOSs derived from the permutated data divided by N was considered as the p-value for the ROCS.</p>
<p>The <strong>parameter</strong> <code>siteCoFile</code> is file name with full path for storing the RCOS between all pairs of columns, and the related p-values. It’s ‘siteCooccurr’ by default.</p>
<p>The <strong>parameter</strong> <code>sampleTimes</code> is an integer of permutations in the simulation when calculating the p-values. It should be greater than 100.</p>
<p><strong>In the example listed below, we assign 10 to the parameter <code>sampleTimes</code> only for testing purpose. Normally, It should be greater than 100.</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pairwiseCooccur =<span class="st"> </span><span class="kw">siteco</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>, <span class="dt">conservativeFilter=</span><span class="fl">0.95</span>, <span class="dt">cooccurFilter=</span><span class="fl">0.9</span>, <span class="dt">siteCoFile=</span><span class="st">'siteCooccurr'</span>, <span class="dt">sampleTimes=</span><span class="dv">10</span>)
<span class="co">#&gt; reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating bigram frequency ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......completed. (10,42)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; calculating cooccurrence..........</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 1. creating siteCoFile (siteCooccurr), sampleTimes: 10 ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1] 1 3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1] 4 6</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1]  7 10</span>
<span class="co">#&gt; completed</span>
<span class="co">#&gt; </span>
<span class="kw">readLines</span>(pairwiseCooccur$siteCoFile)[<span class="dv">1</span>:<span class="dv">5</span>]
<span class="co">#&gt; [1] &quot;Site_i Site_j Co-occur p-value&quot; &quot;3 25 0 1&quot;                      </span>
<span class="co">#&gt; [3] &quot;3 33 0 1&quot;                       &quot;3 43 0 1&quot;                      </span>
<span class="co">#&gt; [5] &quot;3 45 0 1&quot;</span></code></pre></div>
</div>
<div id="generate-the-co-occurrence-network-network-module-network-property-and-rcos-file-by-one-command" class="section level3">
<h3>Generate the co-occurrence network, network module, network property and RCOS file by one command</h3>
<p>The <strong>parameter</strong> <code>siteCo</code> is a boolean flag to check whether the residue co-occurence file would be calculated.</p>
<p><strong>In the example listed below, we assign 10 to the parameter <code>sampleTimes</code> only for testing purpose. Normally, It should be greater than 100.</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cooccurNetwork  =<span class="st"> </span><span class="kw">coocnet</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>,<span class="dt">conservativeFilter=</span><span class="fl">0.95</span>, <span class="dt">cooccurFilter=</span><span class="fl">0.9</span>, <span class="dt">networkFile=</span><span class="st">'cooccurNetwork'</span>, <span class="dt">module =</span> <span class="ot">TRUE</span>, <span class="dt">moduleFile=</span><span class="st">'cooccurNetworkModule'</span>, <span class="dt">property =</span> <span class="ot">TRUE</span>, <span class="dt">propertyFile=</span><span class="st">'cooccurNetworkProperty'</span>, <span class="dt">siteCo=</span><span class="ot">TRUE</span>, <span class="dt">siteCoFile=</span><span class="st">'siteCooccurr'</span>, <span class="dt">sampleTimes=</span><span class="dv">10</span> )
<span class="co">#&gt; reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating bigram frequency ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......completed. (10,42)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 1. calculating networks ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2. creating network file (cooccurNetwork) ......completed</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 3. creating module file (cooccurNetworkModule) ......completed</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 4. creating property file (cooccurNetworkProperty) ......completed</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; calculating cooccurrence..........</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 5. creating siteCoFile (siteCooccurr), sampleTimes: 10 ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1] 1 3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1] 4 6</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1]  7 10</span>
<span class="co">#&gt; completed</span>
<span class="co">#&gt; </span>
<span class="kw">readLines</span>(cooccurNetwork$networkFile)
<span class="co">#&gt;  [1] &quot;EPI823725 3-144 3-159 3-311 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                </span>
<span class="co">#&gt;  [2] &quot;EPI775768 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-326 186-219 186-278 208-238 208-261 219-278 238-261 267-312&quot;                                                                                  </span>
<span class="co">#&gt;  [3] &quot;EPI814365 3-144 3-159 3-311 25-121 25-216 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 142-197 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                          </span>
<span class="co">#&gt;  [4] &quot;EPI787403 3-144 3-159 3-311 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 142-197 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                        </span>
<span class="co">#&gt;  [5] &quot;EPI787049 25-208 25-238 25-261 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-326 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                     </span>
<span class="co">#&gt;  [6] &quot;EPI775712 3-140 3-158 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-225 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                              </span>
<span class="co">#&gt;  [7] &quot;EPI777683 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 159-225 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                                                  </span>
<span class="co">#&gt;  [8] &quot;EPI787395 3-144 3-159 3-311 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 144-159 144-311 159-311 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                </span>
<span class="co">#&gt;  [9] &quot;EPI731085 43-45 43-48 43-186 43-219 43-278 45-48 45-186 45-219 45-278 48-186 48-219 48-278 53-114 53-203 62-259 92-194 114-203 121-216 128-142 140-158 159-326 186-219 186-278 208-238 208-261 219-278 238-261&quot;                                                                                          </span>
<span class="co">#&gt; [10] &quot;EPI748184 43-45 43-48 43-144 43-186 43-219 43-267 43-278 45-48 45-144 45-186 45-219 45-267 45-278 48-144 48-186 48-219 48-267 48-278 53-114 53-203 62-259 92-194 114-203 121-216 140-158 144-186 144-219 144-267 144-278 159-225 186-219 186-267 186-278 208-238 208-261 219-267 219-278 238-261 267-278&quot;</span>
<span class="kw">readLines</span>(cooccurNetwork$moduleFile)
<span class="co">#&gt;  [1] &quot;EPI823725 (3,144,159,311) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (140,158) (208,238,261)&quot;              </span>
<span class="co">#&gt;  [2] &quot;EPI775768 (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (128,142) (140,158) (159,326) (208,238,261) (267,312)&quot;</span>
<span class="co">#&gt;  [3] &quot;EPI814365 (3,144,159,311) (25,121,216) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (140,158) (142,197) (208,238,261)&quot; </span>
<span class="co">#&gt;  [4] &quot;EPI787403 (3,144,159,311) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (140,158) (142,197) (208,238,261)&quot;    </span>
<span class="co">#&gt;  [5] &quot;EPI787049 (25,208,238,261) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (128,142) (140,158) (159,326)&quot;       </span>
<span class="co">#&gt;  [6] &quot;EPI775712 (3,140,158) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (128,142) (159,225) (208,238,261)&quot;        </span>
<span class="co">#&gt;  [7] &quot;EPI777683 (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (140,158) (159,225) (208,238,261)&quot;                    </span>
<span class="co">#&gt;  [8] &quot;EPI787395 (3,144,159,311) (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (140,158) (208,238,261)&quot;              </span>
<span class="co">#&gt;  [9] &quot;EPI731085 (43,45,48,186,219,278) (53,114,203) (62,259) (92,194) (121,216) (128,142) (140,158) (159,326) (208,238,261)&quot;          </span>
<span class="co">#&gt; [10] &quot;EPI748184 (43,45,48,144,186,219,267,278) (53,114,203) (62,259) (92,194) (121,216) (140,158) (159,225) (208,238,261)&quot;</span>
<span class="kw">readLines</span>(cooccurNetwork$propertyFile)
<span class="co">#&gt;  [1] &quot;name  Connectivity  Diameter  Radius  ConnectionEffcient&quot;</span>
<span class="co">#&gt;  [2] &quot;EPI823725 0 1 1 1&quot;                                       </span>
<span class="co">#&gt;  [3] &quot;EPI775768 0 1 1 1&quot;                                       </span>
<span class="co">#&gt;  [4] &quot;EPI814365 0 1 1 1&quot;                                       </span>
<span class="co">#&gt;  [5] &quot;EPI787403 0 1 1 1&quot;                                       </span>
<span class="co">#&gt;  [6] &quot;EPI787049 0 1 1 1&quot;                                       </span>
<span class="co">#&gt;  [7] &quot;EPI775712 0 1 1 1&quot;                                       </span>
<span class="co">#&gt;  [8] &quot;EPI777683 0 1 1 1&quot;                                       </span>
<span class="co">#&gt;  [9] &quot;EPI787395 0 1 1 1&quot;                                       </span>
<span class="co">#&gt; [10] &quot;EPI731085 0 1 1 1&quot;                                       </span>
<span class="co">#&gt; [11] &quot;EPI748184 0 1 1 1&quot;</span>
<span class="kw">readLines</span>(cooccurNetwork$siteCoFile)[<span class="dv">1</span>:<span class="dv">5</span>]
<span class="co">#&gt; [1] &quot;Site_i Site_j Co-occur p-value&quot; &quot;3 25 0 1&quot;                      </span>
<span class="co">#&gt; [3] &quot;3 33 0 1&quot;                       &quot;3 43 0 1&quot;                      </span>
<span class="co">#&gt; [5] &quot;3 45 0 1&quot;</span></code></pre></div>
</div>
<div id="read-and-plot-network-file" class="section level3">
<h3>Read and Plot network file</h3>
<div id="read-and-plot-a-specified-network-the-package-igraph-must-be-installed-and-loaded-firstly" class="section level4">
<h4>Read and Plot a specified network (The package “igraph” must be installed and loaded firstly)</h4>
<p>The <strong>parameter</strong> <code>networkFile</code> is file name with full path for storing the co-occurrence network for each row.</p>
<p>The <strong>parameter</strong> <code>networkName</code> is a vector of network names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cooccurNetwork =<span class="st"> </span><span class="kw">coocnet</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>),<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>)
<span class="co">#&gt; reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating bigram frequency ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......completed. (10,42)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 1. calculating networks ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2. creating network file (cooccurNetwork) ......completed</span>
<span class="co">#&gt; </span>
network_igraph =<span class="st"> </span><span class="kw">toigraph</span>(<span class="dt">networkFile=</span>cooccurNetwork$networkFile, <span class="dt">networkName=</span><span class="kw">c</span>(<span class="st">&quot;EPI823725&quot;</span>))
<span class="co">#&gt; </span>
<span class="kw">plot</span>(network_igraph[[<span class="dv">1</span>]])</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAMAAABKCk6nAAAA5FBMVEUAAAAAACAAADgAAE4AAGMAAHcAAIsAAJUAAJ8AALMAJE4AOscAPzgAZto0AGM0AHc0AIs0JE40Pzg0WiA0Wk46AIs6AJ86Osc6kO1EAItEealcAGNcAHdcAItcJE5cPzhccgBfkalmAItmALNmtv95RIt5qamBJGOBJIuBWneBcgCBiQCQOouQOrOQ2/+kP2OkP3ekP4uknwCknzipeZupqaSpqam2Zou2///FWnfFnwDbkJ/b///mcjjmck7mcmPmiSDmiU7mnwDmnyDmnzj/trP/tsf/28f/2+3//9r//+3///99UyoCAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAUB0lEQVR4nO2da4PbxBWGh6ztwrJA0iX0Ql1MoGRpaMCmVfBCq+7YIQ3+//+nmqskS77G0sy8ep8P8e7apdI8HmlGc84ZsSHQiNAHQLqFgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGJyhCpaiQIY+ih4YpmAlt6B42f0ZkK/AIAVru4Zd53/MVyAJUj/+c6j43WX4iK9AIiR++OdQ89su8IivQCqkffTnsOW3TeARX4FkSPrgz+G3yfW98ZbrH5aTyddbHzniK5AOKR/7ObydOsHrqfohHy0eTeuG3Vfg4fFkclO8zieTjwId7QUYmuCNWFrB339b/LC6m0k5+vCXygf8V6Do25Pi3fn1vZina3iwgvOZ+uHhkxdSXn32suUT6y/092B1d7v9FUiKoQou9Kkfiiu0lI+efd3yiVz330Jw8dv4j6EO950ZquDvFxXBf/pzVsF8YnU3Mbfgh8c366tkO/DwBMux0vfwd+kFi4enbT1YO57prjxKuJUSPvQz0YKLkbFipu7B4k39Hmy+App50YVXz3+cjH4KdbTvzEAF61GyHUWLn7eGUKXgvBA8v5XiFQdZ6fDqqiJY5tfi7bOtJx0VwbNiulx8Bd5+lWwXHprgt88mk/GLUrAcN59k+a/AwxPVx6/F5jV7cFLsXSuyXwH1HEt/A1bFvTpdv8MUfHi1l8uFqXMoXoML/vAwZIekAAWDk5ZgkMtmn6QkGGbg0ycJtRXO1KVP0mkqoEjHPkmmpZAiHfsklYaCinTsk0TayUU6zieTWyM44TCpXklDsIt0nN/I9VQbXk+uKfgY0hBsw2gePl5IE0Ylv/824TipPklKsHb78Himg14p+CjSEqyW6pVgFfRKwUeRlGCdbaIsq6BXCj6KRATbSEcVZjEZLVTQ65iDrKNIRHAZCKdSSWzQ63YsFWkhOcEuKn3MefBRpCLYRjqu7vQkSTaDmUk7aQguIx1v3YNKCj6ONAQrMi4XnkMyTZVxwf8sUmmrTP/LkJ2TSURwFvoAkiUNwfR7NkkITsNvnPePSAXXGisJv7GOACM7HEO9sdLwG+scLq6jMdQbKzW/kRmO6mAM9cZKz29chmM6FkPEjbWLmEM+IzoUQ8yNtQt9zMvJaCHXU10+re2YAw2yo2s+4WNj9zRWXEihDlYtc62eLwrRM9kUGWyQHVvzFW1jY2N3N1ZsaL+qKJ78zwu1onnT/FaGG2THJlj42NidjRUd6qJjl6kV8+YxBxxkx9Z6ohIb295Y8SGKDvzpdGId6wrFYusD4UYVsbWeKGNj2xsrQopjnjxZuGATVVyrfsxBx42xtZ4oY2NbGytGpNBXnVyXLpU/FEcupCpba95V0Sj6jUoF+R6rE8fWekXb2NjYsrFCH9NBjGBV9LCYLc18F9XFiX/7x0IstWFXQf5GPpr2Zzg2weZ6psus1xoraoR2q6vH57eqVnHlmP/5pTBjRVtBXo8hRx/0VvsyuuYzDw1MzfXtxooWsSy6sBoQ5ipee9QYYtm3dPVp1dmvnvYW0x1f8wkXG9vWWLEi5vr+utQB+Y+23tNjRVdBXo8uBi34jc3xbm2saNnzJEOYsaKtIK/HkKP3X7b+V7o4sL7+j04g2rXVfex+FinFDy8qz0D0GHLA92BFrNER+9m5mjCeue9rrh9oyve+G/Ao2hBnfNN5vH7PbAIidQV5RZ95VZEKBuL1xE/rzb141eMFmoI752c9VrzxFeSL11532aLgXuByITpc8IeHITukCyhYgTQr24KCU32uciSI53QiST4ZPRrAUzqReLNOLgLeGZ1IgokUJwF3QieSYiLFSaCdz4m4QuOrO1doDa3Q+DAE75oGuULja/sKWGh8CIL3TIOEq4zoXuEKjQ9A8L5p0LZgvELj+IL3ToO2BAMWGocXvH8atCUYsNA4uuC906Ask+OqYMRC49iCfV7Q0ryqnHIbkWyTh2qCEQuNQwv2eUH5aKGyS9afL+R49FNWZobVBSvQCo1DC/71pckL0lmo+fX9f9XGHuMvqx+xhcbl3AqGKzQOLdjlBem0MP1P8VLNlbCFxvWlWyes4hUahxesem8lu/PhSXMujPsgejMAwSoWuRRcDLZGjaBkLvini8kLqpb9GLeFJTNkJ1l0XlD1HtyeF5REwcSzABds8oL0KNpOhMSrtmkuBaeJywsqpkhmjCXFmz+0jJJx/WILdnlB7kmWqhnQOgui4JSpTYPaVVJw0tSmQVmbTApOnENbQFAwFE2dFIzFtk9gv8MUvG2UguHI9vyGxUAFO6dm9EXBgCip0MtIFuBTO0AGvhBswT2zg4DnjVpgT+wg6HmjFtTzOgh83qgF9LQOI/wmCslswHUWoKd1ELVbmdlEQQVLq4LtDNmBQvhNFHSw9F9eoHZhzLM6jPCbKOhg6d/D3oQxz+owwm+iIF2ZX8ymwDyrw9jde9yGRiquFrMpMM/qMNJMk+amCy8ntx0OsoJGXQ9VsJ0H200U1EZcXbVE4AfeQxdsh1nzWVeCQz/wHqxg3fJ2QyP12r3fMIaHK/jNxGyikOvXPvzuMNztLXq4gvu4eh7xwLvrW/SQBXfduL5CSFkocby9pU7nX7JBC+728ugrhPhCievpaFH/TPe36IEL7hJXIUT6MmvTmy2JPaxJU3CHCPckxQieqzWrrfc7N0zBHWIqhDjB6+mn01oBid98qVt3q+6gyh4Fd4jfOXhp5mNPFqvJ775x6W+2lvHq+UK6W/XV/eWP4eL/ReLxOwcbweoKnas6epnBFMr8zwtZ3qovr4OCu8TtHFwKfvusrCDhK6GWt2oKTgq/c7C9B8+kePNZubl7Kdjfqik4Jcqdg02hxOVIbF59VL5fCva3agpOCF8hxBdKLH6t+N34WsbS3ao5yEqPfc8ifalbd6vmNClB9jzwtoLzW3erpuAk2fnA29Qyzv2tur3G07tBwcGwtYyX/lZdvLz/8vD/7jQoOCxcLkSHC/7wMGSHvAMUDA4Fg0PB4FAwOBQMDgWDE4vgXZNB4A1v+iEOwbse5wyh1mDHRNFyux7Ihk69RCCGhtuVvxE89RKBCNptV/7GQGoNdkz4ZtuVvzGUWoMdE7zVbP6G2l7d1MswKZaN1Mvi/bZNB8kBQgu2+Rs6sFRHf5sUy0bqZfHeoykNn05owTY4WJW5UTsMuhTL7dTLh48XUo4+aGz9Sw4RjeDi3+VtJcWynnqpdwC+etq2cSjZSySC5cPjm/Vf722K5TcqM6uWepmrwHEKPoNYBMtcR46aFEt1Ka6nXq6nxb+jy8cc4hNcsM3fWD3/cao2+vUplvXUS128e7uACTmC4IJteP/8Vo+kKymWtdRLRfvm7GQ/kQjW+3OvP1+UKZb11Es9P+4g7h+f8IJ1/oYeRatpkkqx/NtHzdRLueIF+ixCC7b5G7o8tzapUiyzbDv1srgF8/p8FqEFa5rLgpmuVMLlwncnjoZrLOxn1rD/O/L+kZ0Sh+BNZkNzjEdTaCir/H2T0fB5xCK4+up/0Vaz+ifIacQhOKv+VHZWbbh0zwC8M4hOcPVarPW63xmAdxZhW8vfYf1ftvyWf+CI+jxCNlazT2ZZ/Wq98V2aAXhnErCtWma/m8q42QnW2hmAdy7hmqrZJ1tnvs3+S8OnEKylGn1yx7OrUnBuViWQN/vtgFAt1d4nW+60SrDeh06v+YNv9tsBQQXnZilYryqITUuv3tjrtvrt+2+LT4Nv9tsBgRqq7JMuYLbok6292gnOZ25hGHiz3w4I1FBln/QBs6Laq1WEzk1V8PoLt/KPvNlvBwQUbPqkD5gVlV6t6/vNij/Y2vfFl2FhBUNv9tsB4QS7PmkDZqUoe7V6z4+tC6RQwTvuEt3tZr9ohBPs+6QJmDWC7Z02t4lJwj6oFHMd4eGylbrb7BePYIMs3ydtwKy+QttereJ3dCqasI+zdOf20XfdbfYLSLBpku+TNmBW1Hq1cjwrBWvDfhTd2Wa/iASdBy91tpkJmBWb+p3WTp2y8vPqnU43+4Uk6KNKP4rO9XYjtTutzPU0Kat+vv4QM1IiC0sIuthgEwuLTin8kpHvwfmsJjhLY8E/uqOMaLkwa7nTln430fWNNuK7zsS04K+faCjB6jmWudOW8XehDvIkIgxLiCNkx5E1e/Wm8hI7MYYlRHEQJdU7bWk/Tb9xGI7hGKpkLmSnev1OQnCjLlAn+z2fTGyCN9VnV7YfpOH3q4VQoQi+LlA3+z2fTHyC9YAqwtHKAf73izChCJVVkQ42Ez2ZCA6hSRbjaOUgwoQiUPBhohytHETYqjEUfAA3WtGb+qnk8Hl9291YGduSBBS8H1fFUAXI6qx/tdQ0T8CwFCoUoSqYg6w2XBXDf6vekN+aoK1RCgVYhD7UquAIDjo+wZUqhnrvcz2tHKdQoUOoUAQKPoyvYih1sJYK2oqhqQ5zZUZZti5QJ/s9n0ycgt14NNeXvHwyivEw66i6Tx+auFAzNOxmv+eTibHlfBVDqV9Xz3+cjNKokbW1WFJZAwu21BmjYF/FUF+hVdCWeBXD1e4gWWMJ1EWFhgsDiFGwr2Kor9C6uuHbrxLowsZmva9uZ7/23t4RCi6rGJor9N212LyOtAfXbNbCT/zPWdgH6/EJ9lUM7RVartTwJUq/tStvo7qI/1TQx67xCTbEF9zUZGtEVXkna/9UiPOJt/ViC09sUL/y1jpt5Re3bZDLm+z7qVy0zRd9EOW+K28p2G0b5PMmi/nxNQUnwN4rb6UHuz0pbN6keu31sRwFn4mw3dIlYhwS7CoUFK8UnAA6W93XhNGCq+G/5Y9WsMuGVq8UnACmA/uaMFtdODMVr3V1AiPY5U2qVwpOgPIWbAOx7MNnj/ug2TbI5U2q1zEHWQngBbuFr7IcwaY+D9aCXd6kfe1xBzcKPg8neOkCsbYv0Z6xmwe74XS/2wNR8HlI34VtINauQZbZNqgiWPxMwSngBdtArB3TJLdtkBcsNhScBOUoywRi7XqUVf8slwvTwVmzo6x6Q249meaCf4oIaWvCtPTMxtISQ3YSZPdy4c61w/6h4HegfuVtX/EPXXyCgt+J9pCd6l9DJzdT8AWxG8TUAnkCHxIFX5TGthPB/VLwZcmyoCGULURwCFBEV5sg/BFAEV9tguAHAIULoTRbTuhVwtB5rxR8QXwIpdlyQmU5P5oGNjxMwV09OCwDsHQ//ngh5eiDsFlVQxTc3aN/4Xah0KGWuQoFuHraY/hG2yEF/X8PQoeLd8JXwFa34FwtBFNw33Q5UfVR7nrLCZ3LMAqc5T84wZ1OVKUPwNJlZNRgesJ7cK90PFEtBeemTtB733EU3SO+5O/Svl66iF5FsAnk6TeEsoVBCXZF9NT4Vm/nc/Eiej6E0gTyrEJfoAcm2BXR02W48uv7SxfRsyGUZssJfQsO/RxrYIJdEb2HT4r5S/FPJ0X0IqtNEMMx9IgpopebyuyzboroxVWbIIqD6A9TRM8L7qiIXky1CQYm2BTR04IfHs9SKqJ3LgMTbIrouXtwQkX0zmZggk0RPT2KXl7fp1NE73yGJdgV0cuN3KiL6F2IQQkui+jZJ1nxFtG7GIMSbIhsotoxAzjFLYLm+vXPEM6xRlbZiTh8WHr3DE3wzhxAVAYmONvzGybDEpzt/RWSQQlu5OoOwDC64P2puhScOAdTdfENQws+nKpLwSmzp+a+g4IT5pgIaApOl6MioCk4VXwEtFo5UgEcq7vW4Ct4w6CCfQS0XE91ud/1dNK+H3dM8VNdACrYRUAXYu1W03J53XKu+AtLuCdm8r/mrlx3q+ABLA2jnpeNgF5PP53akuwtgmMredQFoKflI6AnTxarO224KTi6kkddgHlWm1oEdK5HW8ur+6y22hBfyaMugDwpzdiXydA5DHa7Ir/ljZtH+b3LLp1JGgmwgk0EtHar49yr+1EVjv9l51F+77KLZ5JGAqpgFwGt5M31TKkeH/vrl2Ye5fYuu3QmaTSACi4joOd2N/Hi13o5FDuPkmbvsk4ySWMAVLBh7zTXzKOkS8bvJJM0AqAF731QJaxZt3dZN5mkwUE8pyq7HzWbeZRZjSjuv6iZpOiC9zC2q01mgIWaSTpcwWYepQ3PZxI2k3Swgt08yo6yYDNJhyrYzaP83mWomaRDFWzgciE6XPCHhyE7JGkoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHD+D2nbGAW4R4hZAAAAAElFTkSuQmCC" style="display: block; margin: auto;" /></p>
</div>
<div id="read-all-networks-and-plot-one" class="section level4">
<h4>Read all networks and plot one</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cooccurNetwork =<span class="st"> </span><span class="kw">coocnet</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>),<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>)
<span class="co">#&gt; reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating bigram frequency ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......completed. (10,42)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 1. calculating networks ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2. creating network file (cooccurNetwork) ......completed</span>
<span class="co">#&gt; </span>
network_igraph =<span class="st"> </span><span class="kw">toigraph</span>(<span class="dt">networkFile=</span>cooccurNetwork$networkFile, <span class="dt">networkName=</span>cooccurNetwork$xnames)
<span class="co">#&gt; </span>
<span class="kw">plot</span>(network_igraph[[<span class="dv">2</span>]])  </code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAMAAABKCk6nAAAA6lBMVEUAAAAAACAAADgAAE4AAGMAAHcAAIsAAJUAAJ8AALMAJE4AJpsAOscAPzgAZto0AGM0AHc0AIs0JE40Pzg0WiA0Wk46AIs6AJ86Osc6kO1EAItcAGNcAHdcAItcJE5cPzhccgBmAItmALNmOsdmtv95qamBJGOBJIuBWneBcgCBiQCQOouQOrOQtv+Q2/+RX5GRqamkP2OkP3ekP4uknwCknzipqZ+pqaSpqam2Zou2///FWnfFnwDbkJ/b///mcjjmck7mcmPmiSDmiU7mnwDmnyDmnzj/trP/tsf/28f/2+3//9r//+3///8EhFGUAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAVlElEQVR4nO2dDXfbxpWGx5aIZt2obRyladpoy3i3UaJk3TpkUnhpp2UFUnFl/f+/U8wnQJCULAKYufPyfc5JZEs65gAPifm4d+6oOwKNSt0AMi4UDA4Fg0PB4FAwOBQMDgWDQ8HgUDA4FAwOBYNDweBQMDgUDA4Fg0PB4FAwOBQMDgWDQ8HgUDA4FAwOBYNDweBQMDgUDA4Fg0PB4FAwOBQMDgWDQ8HgUDA4FAwOBYNDweBQMDgUDA4Fg0PB4FAwOBQMDgWDQ8HgUDA4FAwOBYNDweBQMDgUDA4Fg0PB4FAwOBQMDgWDQ8HgUDA4FAwOBYNDweBQMDgUDA4Fg0PB4FAwOBQMDgWDQ8HgUDA4FAwOBYNDweBQMDgUDA4Fg0PB4FAwOBQMDgWDQ8HgUDA4FAwOBYNDweBQMDgUDA4Fg0PB4Byp4ErVVKlbEQNQwfcL1D+rqb/EbFMaIC/xAYHmhxbIy98A8QofENj68REYBrzABwRu/BjfMN71PSCw82N4w3CX94DA98WzN/q7158UxVn9dVYUXyRqaRzABN9+VRTT2tv6wn6tmRT/9f/Nzy+c4EVhfmF2Vj29gDaMJfj2xVwtTufV6ut5rdCaPp23f0MtjOD1n43869/Oq+q09QbAA0vwv39W1frLq+qfV1W1ujzTfs82n9FO8NJ+wJf1m6E6+f3/pmpvBLAE6y74+lPX+daP32pWG1R3ZYMVvLosTBe8nFxRcF6o6+dvrN/V5bT+AH92UZz+pf1z+wk2jvXP67+dfvQyWXPHB03wpCj0Y1f3r7XpZfF8vmoPsu6qiRNsP+F6MF2wD86ISi2Kc+Pvhyvbx6pl0X4EN4KXZ+bLk+84is4Jtbo0ghd+EKVuv9oj2M6jJh//nKyxEcATXM2M2try9bfri2ml3n3e7mNfnzjBtrNeYT+gAQXfnWhx9TTIdMaLU3X3+jfND/VCSD1y1l2v/iTXX3s+n8UHlrEE39RaP1Z2ncouRdZfftP9reHChRkEluW27HAeFDiUlxwCy2Ib1oeHBQ7yZM0isCy1XT2J0TXmEVgW2qwMyCSwLLNV8vFxSfGBZQo+iPffz5WNR0oPLFPwQfzyUrl4pPTAMgUfiLLRCvGBZQo+EGXikfIDyxR8IMotZksPLFPwgVRKxyPlB5Yp+FAmUz8BFh1YpuADuXmi45HyA8sUfBg3Lh4pPrBMwQfx1sUjhwssjwUF94DhQnQY8AenZMoONGXqBnwAFNwDCsYmB78UfDglBWOThV9xgsWPSkMD8/ArTLD4eWXTwDwe0MIEi18ZEt/AbSS1U3wiufgG7kBQM8Unkotv4C7ktFJ8Irn4Bu5ETCNdhbJQ4Gp1eSIsfm4EL3QQ2DVSzL27DymNdBXKQoGr+h6evEndqA0qZatuNY0UPZ3zSBHsyt+EAlf1Z+WZmLYZlK26VbUaKauBuxHTxlDfyGYpihQ8cwV8fCNlNXA3YtrYCNYJ5SIF26pb86aRshq4GzFtbAS7hHJ5gk3VrctWpp2sBu5GTBubCmUuoXwhbpBlNiAt7SBfN5KDrEfhBS9cQvlC3DTJCNaVmVwj5dy7e5DTSCfYFrgSKdi4vf7dlW+knHt3D3IaaSuULUNCeSFuo4ApRe32jNaNlHPr7kNKK12FslDgal1/EbdZT81M22wjn6ZuzYchRbBBfDROfAO3EdXOMqOAf+qWfCiSGqpzJPqk7ERJ9xGfU9RBkOCeOTD5fbiiIOhu9BOcYfcYBUE3o5fgHLNpoiDmXvTLUswymyYKYm7FkH5puCHtnWiGpL38+gMJq2pp/rAoCoEVq9KQUnB73NtHcDiQ0Jar0mXnnl6INJxgjpVQcHvc23ME7WPJr76p/2Bi8afiVrITTeTSCR5w3OsFL6f6Dybec/K5tIXsRBO5ZIKHHPc2R4rqP5iw7dOvpD2jE03kUgkedNzrBL+a3yc48RpjqolcGsGhXvrq0o9/Jz0Kidl0H50m4AWr607l19QLmckmckmuN9RLX/vxb/cc50diBM9MmLaY6j64c9xZ6oXM8IZ2WyNMRDlOuDuJ4FAv3X7kdpzj/EhCwt7CjaLVX3/9t9bPEy9kNgcA2K0ReutLESljJd0Ta9YSbM9xPpxwIKH555bPzIGUZVj9TL2Q2RwAYLdG/EOnJJ3GqX2YTHBIb39TuYzyg7tgl+7TvF8mfiXLOhawkKmaA8k9kzjzuGSCQ3q7OUqye47zIf/ibn9l+XfXz7cOwNk6znBsVOtAcut4HSlpNJHgUC994dYWu+c4P5594+SwkOkPwHn2Rs1iG1bNgeRua8TySaRXjvMyW/h66Y3g295LE3tmus0yiMacHx19IdO+odtbIyZXcV45keBQL931wVvnOA+IE+wOwDEz70n04s6TcPiO3RoR6wmdSHBTL31m43udc5wHxQr2B+DUXfFZtLsbsG/oZmtEtCd0GsG+XrqZEJrx765znIcixJrMATj6oxx9U4J/Q/utEdHG0GLChWO+TrNv0dzc1dc/9piSHUR4Q1d2a0RVPf2fWM8QKQH/MWkE6wNwZueVep0iWnxk4UJDnAhPS/DUDuhuX6Q4IaVVBzHii8Z7qWSEhUw9F61H0eruJlG+h39DU/CAuIVMfwBOtYq2zr8fCh4aafse4hkWcbkRSB3w7xLNsJDrjYCsbYEUjE4swxScCApGp4zTa1BwKiLVq6DgVESauVFwImIlelJwGqIlelJwEuIlelJwAro7d3Tew1h75Sg4Prcv5qbspd+5o8+AmIxlmILj8++fVbX+8ioklZozIMZKWaLgFKjq+tMqCDZQMBLK7etoBK/4iEZiUrgdLM1BFWNl8lJwCiq1KM43BI+20YGCk6DMDppG8ILzYCxUNZu2BC/P1d27/xvnlUb5V8lDnNhR1sxvmyp67p7dCwXHR29k+VhVYeeOPQOC0yQwGC4UxQjJFwz4y2EkF0zZEYK0tPnHkF2DE5D1cQG5tTcBqats9SOz5iZAQJWtPuTV2hQosy//3E1bMzn3uyGv1iagUrrww/ri3CRe6IqTcjY4fQgU/ACquv7t3FRAMokXuuJkVvcsq8amQNnqVtef2MptMwoGoxasCz05waaEalb3LKvGpkDZs3qWtpytSbXJ6p5l1dgU1GMqXd7DpdjoipMcZGGh7KPZJmBMOQ+GQzWZF8tzXXEyr1uWV2uToIt5mwe0SbyIXueyJ5k1NwXvCpcBaRIvnqZuziOh4A+A4UJwIlVbGIX8WpyASPVSRoGCP4CIpSUHh4IfJme/FPwwWful4AfJ2y8FPwgFY5O5XwreSTMpyt0vBe+gtayRvV8K3qa9MEnBeGS9j2Gb/K9gYPLex7BN9hcwMJnvY9gm9/YPzHt3mLTby1AlOGp4YCi4jT8t3O1l0NtVimcUPCSJw3L2KFq/l6EW/Oqb6GcND4sswckD6yocOW9T3ZfTBQUPR/rUGCfY72VY/7mi4OEQMAO1gsNehlfzoQQn63oECZYwA3Wnhbu9DNffVvU3BhCcsOuRI1jGDDQcJq33MsxMomz/Qr8pux4pgrvHGOib+0WCdgTBvkzopP88OGnXI0Tw++/nSu+e98cY6Jno04sEht1p4W4vQ+3jbW/BabseIYJ/eans7nn30TEz0dOR6nPupzkt/Nzb6C04cdcjRLCtddIINjPRk9+PddbMQy0Z0IVf/KyvyQ7QxztBZyeCBJvd816wnoluCI44zxhyzOsXP93ca9QTdHYiSLA7qKKZiZ5+9DL8NO48Y8B3k/KjtVff1H8Y9QSd3a8f8bXupVI/XLWGr2YmGvrg9EtcB+MFL6fhfIbjFHw3mW7OT6on3/lRtIAlroNxgvWip7uw0U7Q2f36EV/rXm6e6N3zbcGTX//NpkRJWOI6GCdYL3q6CxvtBJ3drx/xte7jpvCFTuwxBvX7XD+gy7JMPc/oiV381G/dsHQy1gk6OxFys96aNcGzcIxB3QW75/NPf7RLXGZ/vR5bz0Y73mAcjGC36GkupHvLx50fCBFsKHeMpfwSlxZf6BlH/QFXs5wMdxY/l+eqbJ+gM/b8IK3gjTdv2b7a0mUk+yWuf5gqKOe2nlFWeVJu8dNP8PX79FfNCt3o84OUgjffvNZoUB5SzsMSlylDZmIRkxRhiMNwi59esDtBp/Tv3/HnBwkFb755g9DS/bkMv+aWuKr1f+vhyidn69ySLHZ9TI3iCPODdII337zNHpGy9Z/9PX8W69JEAJbZlara09GWZYz5QbJ7tffNGyqeGMtlGZa4TKHIavX1j8Vp7ChTf3YNlV0cwhWSH2l+kCx9ce+bt+yOtPwSl3lCV7PzSr3OaZC1FxeH8IXkR5ofJBUcAmjtIszdDisscS3tuQnTSt2+yO8jvAO7yOULyY80P0gkuLJnM9YXuP5T/QY+nbsivZ1+SbWXuOwT+vKZuruB+AQ3gSYzTxhpfpAq/1hflQmg/Ut/iL/UZZjLHeuSKixxuSd0tdJHd0L4bQk284Rx5gcJBTcBtOtP/fN4u2suO9/ObgS9n0awnSeMMj9IJ7gJoNnLawQvQ0T4rHJrIMn3tIxBFVYxR5wfpBMcAmgL28eqst012+iCPaTIfIizLRZ5D17wws6SRpkfJBtktQNoi6K+tqp0IyzTNesPuHsk+xWP/OtldHGCbSH5keYHyaZJrQCamSGEJ7TrmpfuZ6q9hgmGjUPYQvLzkeYHSefBYZAxm1qRTde8unSD540lLihcHGLh5wnjzA+SLlWGUfTzN8osS7a7Zu14iizYABwuVFbw0sTxXTM2u2Y9/bdL0poSUjBywH/Xm3eja66WZ9BjLAduyk4rqhC+tdE1L6cuUow4R4qEjJSd5tO53TULqNyRM0LuWTMXcl2zXsequ2bd8WKuU8ZCyC1rTXY7mTxb4SXyKKTcsdZQavOJnPW2BgFIuWFl86VsuubctzUIQMz9KjdGXP4D/femdqRLnZ0U0ff9Z40YwRsPZu831I7UWxpM1P/idJ6ylfkhRfDG0Ko1azJTJpOupKPE64szPqMfh5DbtS9J2gvWmwJqyTOdvJWwmRki43btHSorv9//TKdkrS8+u8gxKTolIgTvHyqHAggm6WNZPJ+vOMh6FHIE20wsO1x2rSpLl7a0+vrHenRliiupZdwyRLkjQXArE8sNl+vZUml34FnBs3PzIyP49isKfgQSBDeZWGG4HMZZE7v5QRc5/NPcpC29+/zlvf8a2UCKYJuJFYbLoVkmbcl8W3tfnKq71zlt70+PEME+SdoNl71gv31aZ2iZZ3WRWYGO9AgRHDKx7HB5Y7LLcGEfJNyyViaWGy5vZm8w4N8DEfesycRyw+W4lYagkSFYf0AXZq3ZDpdFtAoDGbfSZ2K54bKMRmEg5F4q3wfr4bKQNmEg5WZyqDwSYu4mh8rjIOh2cqg8BoIEkzGgYHAo+EBy6VAo+CDyGRKKb6BIMprUSW+fSHLaLiW8eSLJaruU7NaJJK/tUqIbJxPlq/BtFcqViOjGiUTngNoqfJuFcoVCwY9FhSp8TaHc7d8SM02m4Meimip8VbtQ7sbvyJkmp29BbqimCt9GodyNX5EzAEvegOww9mwVvlAod8dvSDGc+vXzo7L+3GFdtlDuxi/ImiZT8KOxApdWsC2UW5blZi03OYYp+NE4wW6YNZs6g06yOwzJ5g8a0p60SMGPR4XRVVONz1KWP12E066c4HXxjILz4p096HazUK4nnLThv776JulRixR8CPfMg7qCl9MFBefH/pWMjmC9bZKCc2TfWmRHsN42ScHZsqNIeTgMye53/raqv0HB2bLD8IZgt20yYVURCu7FQ4I1E86D4zF0FG+HYHsYki0YZAZibyk4FiNE8bqGfVURneyhv9bTKAqOxRhRvB0f4Y0i9cnvb/IGxGOUKN62YHOCGwP+8Rknircl2H2DKTuxGSuKV9771/QciWA99jHxvYX7OhuoolrnkyrO75EIfv/9XC202eWpKXhpSjXN+hvu9rXy/B6J4F9e6lS5s/q/qSl2aoqe9g/Ed0fLAv0eiWDjYnZWXf+unprW/zPpFpMv+v+bG4Yl+j0iwfrTqwtOm2pruuhp3yBPd1gu0u8RCdZJNkGwLnra89KFJdftQ2arhqdSP1xVVvD1J1Nd9LTf6R7dYblemhRZif5YBN9NtAbfB+uip+p1j0FWd1iu96FNRJ4HcyyCb57o6LsZRS+evTFnA9y+OFxId1hu9qFNJJ41cCSCb3TYXZ/bYuXWo2h1d9NrmrQ5LNcP6+ungzV3QI5D8FuTV3HWdJmr+q/9psGdYbnOkBZ5L0U2agQGP0q8OyzX+9DYByfDzlGHjOJ1huWVPvi278rJGByH4LAGMWAUb3NYXvPkOwqORFfjGCG9zWG5eTC8ljgRBhS89SDuCh1CcGdYrl/v3R9S5l7tA0/wVohnX9JFH7rD8mXvUflYwAn+gBDPYFEBScl1+xDbsAP5kBDPcGEfQcl1+5DbsoP4oBDPkHE9Mcl1+4ASHEI8umvUM9TV5c6gr8zI7TggCQ4hnmp9YY64XF8UJ292/CIF54kP8dRiz9wTerHzlC0KzhUb4qlm5vNLwRo0wXppaX3x2YU9hni3YPEDoyFBE2xCPMXz+erSGN4hOIOpzZBgXWUrxLO0xSS3Blk5LE4MCdhFmhBPKwq/OPl5c6lSVKHQGGBdow3xGLc2hmcq3OwvJIl19TuBukQf4tF19m01WB8BcIUkM8llHhKkK2xCPDP7ZV3/9aOQ6lj+9Mcmh3nQLYaSQRJs2TuK8gtd/iyNobYYygZP8N55kF/o8mdpDLTFUDiAgu8pMxgKteuzNIbZYigdSMH7sAtdla/2PMQWQ/EcmWBr1p2lMcAWQ/nAX2Abu9Blw8V1/9t7i2EOHJVgu9ClMQOsvlsMs+CoBNuFLmN4Nq16bzHMgmMS7Be63ChrgC2GGXBEgv1C19KdqTHAFsMMOCLBFoYL0WHAHx6m7BAcKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBwKBgcCgaHgsGhYHAoGBwKBoeCwaFgcCgYHAoGh4LBoWBw/gNDsbLRc1d5FgAAAABJRU5ErkJggg==" style="display: block; margin: auto;" /></p>
</div>
<div id="save-graphics-as-pdf-or-jpeg" class="section level4">
<h4>Save graphics as PDF or jpeg</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">workspace =<span class="st"> </span><span class="kw">getwd</span>()
cooccurNetwork =<span class="st"> </span><span class="kw">coocnet</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>),<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>)
<span class="co">#&gt; reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Calculating bigram frequency ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......completed. (10,42)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 1. calculating networks ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2. creating network file (cooccurNetwork) ......completed</span>
<span class="co">#&gt; </span>
network_igraph =<span class="st"> </span><span class="kw">toigraph</span>(<span class="dt">networkFile=</span>cooccurNetwork$networkFile, <span class="dt">networkName=</span>cooccurNetwork$xnames)
<span class="co">#&gt; </span>
<span class="co">#save as PDF</span>
<span class="kw">pdf</span>(<span class="dt">file=</span><span class="st">&quot;sample1.pdf&quot;</span>)
<span class="kw">plot</span>(network_igraph[[<span class="dv">1</span>]])  
<span class="kw">dev.off</span>()
<span class="co">#&gt; png </span>
<span class="co">#&gt;   2</span>
<span class="co">#save as jpeg</span>
<span class="kw">jpeg</span>(<span class="dt">file=</span><span class="st">&quot;sample1.jpeg&quot;</span>)
<span class="kw">plot</span>(network_igraph[[<span class="dv">2</span>]])  
<span class="kw">dev.off</span>()
<span class="co">#&gt; png </span>
<span class="co">#&gt;   2</span>
workspace
<span class="co">#&gt; [1] &quot;C:/Users/Jerry/AppData/Local/Temp/Rtmpgvbts6/Rbuild146832c62740/cooccurNet/vignettes&quot;</span></code></pre></div>
</div>
</div>
</div>
<div id="outputs" class="section level2">
<h2>Outputs</h2>
<p>The outputs for the package <strong>cooccurNet</strong> are listed as follows:</p>
<ol style="list-style-type: decimal">
<li><p>A <strong>co-occurrence network</strong> (<code>networkFile</code>) was given for each sequence, which could be easily transformed into the format of igraph by the function toigraph();</p></li>
<li><p>The <strong>inherent modules</strong>(<code>moduleFile</code>) in each co-occurrence network and <strong>some basic network attributes</strong>(<code>propertyFile</code>) such as <code>connectivity</code> and <code>clustering coefficient</code> for each network were given;</p></li>
<li><p>The <strong>extent of co-occurrence between residues</strong>(<code>siteCoFile</code>), defined as residue co-occurrence score (RCOS), were given for all pairs of residues.</p></li>
</ol>
</div>
<div id="optimization" class="section level2">
<h2>Optimization</h2>
<p>During the development of <strong>cooccurNet</strong>, we found that <strong>cooccurNet</strong> consumed lots of RAM, which may lead to crash of computer when the data inputted is large. In order to solve this RAM shortage problem, several approaches were taken into account.</p>
<div id="matrix-type" class="section level3">
<h3>Matrix type</h3>
<p>The <strong>parameter</strong> <code>memory</code> indicates the type of matrix used in <strong>cooccurNet</strong>. It could be ‘memory’ or ‘sparse’. If it’s set to be ‘memory’, all data would be manipulated in the RAM by using normal matrix and package ‘bigmemory’. If it’s set to be ‘sparse’, the package “Matrix” would be used to manipulate massive matrices in memory and initialize huge sparse matrix, which could significantly reduce the RAM consumed. In default, it is set to be NULL, so that <strong>cooccurNet</strong> would determine automatically whether all data is manipulated in the RAM or not, according to the size of data inputted and the RAM available for R.</p>
<p><strong>If your system has not enough RAM, you can set <code>memory='sparse'</code> to keep this program working properly.</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cooccurNetwork  =<span class="st"> </span><span class="kw">coocnet</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>,<span class="dt">conservativeFilter=</span><span class="fl">0.95</span>, <span class="dt">cooccurFilter=</span><span class="fl">0.9</span>, <span class="dt">networkFile=</span><span class="st">'cooccurNetwork'</span>, <span class="dt">memory=</span><span class="st">'sparse'</span>)
<span class="co">#&gt; reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; preprocess file......preprocess file......completed. (10,42)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 1. calculating networks ......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2. creating network file (cooccurNetwork) ......completed</span>
<span class="co">#&gt; </span></code></pre></div>
<div id="ram-usage-optimization-for-cooccurnet" class="section level4">
<h4>RAM usage optimization for <strong>cooccurNet</strong></h4>
<table style="width:169%;">
<colgroup>
<col width="38%"></col>
<col width="33%"></col>
<col width="54%"></col>
<col width="43%"></col>
</colgroup>
<thead>
<tr class="header">
<th align="left">Dimension of data inputted</th>
<th align="left">Size of data inputted</th>
<th align="left">RAM requirement without optimization</th>
<th align="left">RAM usage after optimization</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">10 x 10</td>
<td align="left">400 b</td>
<td align="left">1.8 Kb</td>
<td align="left">1.6 Kb</td>
</tr>
<tr class="even">
<td align="left">100 x 100</td>
<td align="left">39 Kb</td>
<td align="left">1933 Kb</td>
<td align="left">20.7 Kb</td>
</tr>
<tr class="odd">
<td align="left">1000x 1000</td>
<td align="left">3.8 Mb</td>
<td align="left">1905 Mb</td>
<td align="left">1.9 Mb</td>
</tr>
<tr class="even">
<td align="left">10000x 10000</td>
<td align="left">381.5 Mb</td>
<td align="left">1862 Gb</td>
<td align="left">190.7 Mb</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="high-performance-computation" class="section level3">
<h3>High performance computation</h3>
<p>The <strong>parameter</strong> <code>parallel</code> is a boolean flag to indicate whether the parallel function is in use or not.</p>
<p>In the current version of <strong>cooccurNet</strong> (0.1.4) , the package <code>parallel</code> from R core is in use. For better memory handling purpose, we use <code>FORK</code>, but it only supports <strong>Unix/Mac (not Windows)</strong> system.</p>
<p><strong>Only if the parameter <code>memory = 'sparse'</code>, ‘parallel’ would take effect.</strong></p>
<p>Parallel uses <code>max(1, detectCores()-1)</code> cores. If you want to specify yours, plesae create a text file named <code>cores</code> in your workspace and put a number in it.</p>
<p><code>cooccurNetwork  = coocnet(dataFile=getexample(dataType=&quot;protein&quot;), dataType=&quot;protein&quot;,conservativeFilter=0.95, cooccurFilter=0.9, networkFile='cooccurNetwork', memory='sparse', parallel=TRUE)</code></p>
</div>
<div id="debug-your-calculation" class="section level3">
<h3>Debug your calculation</h3>
<p>The <strong>parameter</strong> <code>debug</code> is a boolean flag to indicate whether the debug message will be displayed or not, it’s FALSE by default. You can add it into each function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data =<span class="st"> </span><span class="kw">readseq</span>(<span class="dt">dataFile=</span><span class="kw">getexample</span>(<span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>), <span class="dt">dataType=</span><span class="st">&quot;protein&quot;</span>,<span class="dt">debug=</span><span class="ot">TRUE</span>)
<span class="co">#&gt; Reading file......</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     reading sequence from file time cost:0.12 secs;memory applied:93.1 MB</span>
<span class="co">#&gt; character level:A,C,D,E,F,G,H,I,K,L,M,N,P,Q,R,S,T,V,W,X,Y</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     split string to character time cost:0.47 secs;memory applied:93.4 MB</span>
<span class="co">#&gt; completed. (10,329)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     total co-occurrence.readsequence time cost:0.68 secs;memory applied:93.4 MB</span>
<span class="co">#&gt; </span></code></pre></div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
